### PR TITLE
Added PragmaControlledCircuit

### PR DESCRIPTION
**Pull request by nfwvogt**

### CODE REVIEW

File: .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+#[test_case(test_roqoqo_1_4::operations::InputBit::new("input".to_string(), 1, true).into(); "InputBit")]
+#[test_case(test_roqoqo_1_4::operations::PragmaLoop::new(2.0.into(), test_roqoqo_1_4::Circuit::new()).into(); "PragmaLoop")]
+#[test_case(test_roqoqo_1_4::operations::PhaseShiftedControlledPhase::new(0,1, 1.0.into(), 1.0.into()).into(); "PhaseShiftedControlledPhase")]
+// 1.3
+#[test_case(test_roqoqo_1_4::operations::ControlledRotateX::new(0,1, 1.0.into()).into(); "ControlledRotateX")]
+#[test_case(test_roqoqo_1_4::operations::ControlledRotateXY::new(0,1, 1.0.into(), 1.0.into()).into(); "ControlledRotateXY")]
+#[test_case(test_roqoqo_1_4::operations::ControlledControlledPauliZ::new(0,1,2).into(); "ControlledControlledPauliZ")]
+#[test_case(test_roqoqo_1_4::operations::ControlledControlledPhaseShift::new(0,1,2, 1.0.into()).into(); "ControlledControlledPhaseShift")]
+#[test_case(test_roqoqo_1_4::operations::Toffoli::new(0,1,2).into(); "Toffoli")]
+// 1.4 - add this into the next compatibility crate
```
Comment (kbarkhqs) [inline .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs:98]: This comment should be changed to just `1.4`

File: .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#[test_case(test_roqoqo_1_4::operations::GPi::new(0, 0.1.into()).into(); "GPi")]
+#[test_case(test_roqoqo_1_4::operations::GPi2::new(0, 0.1.into()).into(); "GPi2")]
+
```
Comment (kbarkhqs) [inline .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs:101]: I would suggest adding in the operations (commented out) added in version 1.5 here, with a comment above to uncomment them in the next compatibility crate, so it's easier to keep track of

File: qoqo/src/operations/pragma_operations.rs [hunk 3]
```diff
@@ -1210,6 +1210,22 @@ pub struct PragmaConditional {
...
+///
+/// The circuit is applied when the qubit is in state 1.
+/// Note that this is a unitary operation (for example a CNOT(0,1)
+/// is equvalent to a PragmaControlledCircuit(0, [PauliX(1)]) but it cannot be represented
+/// by a unitary operation in qoqo for arbitraty circuits.
+///
+/// Args:
+///     controlling_qubit (int): - The qubit controlling circuit application.
+///     circuit (Circuit): - The circuit executed if the condition is met.
+pub struct PragmaControlledCircuit {
```
Comment (kbarkhqs) [inline qoqo/src/operations/pragma_operations.rs:1224]: The name should be added to the mod.rs of operations: https://github.com/HQSquantumsimulations/qoqo/blob/main/qoqo/src/operations/mod.rs#L134  with a comment about it being in version 1.5

File: roqoqo/tests/integration/operations/pragma_operations.rs [hunk 4]
```diff
@@ -3109,3 +3109,176 @@ fn pragma_change_device_substitute_trait() {
...
+    assert_eq!(pragma.controlling_qubit(), &1_usize);
+    assert_eq!(pragma.circuit(), &circuit.clone());
+
+    // Test InvolveQubits trait
+    let mut qubits: HashSet<usize> = HashSet::new();
+    qubits.insert(0);
+    qubits.insert(1);
+    assert_eq!(pragma.involved_qubits(), InvolvedQubits::Set(qubits));
+}
+
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/pragma_operations.rs:3130]: I would suggest adding a couple of lines testing the `All` and `None` variants of the control circuit's InvolvedQubits

File: roqoqo/tests/integration/operations/pragma_operations.rs [hunk 5]
```diff
@@ -3109,3 +3109,176 @@ fn pragma_change_device_substitute_trait() {
...
+    );
+}
+
+/// Test PragmaConditional Serialization and Deserialization traits (compact)
+#[cfg(feature = "serialize")]
+#[test]
+fn pragma_controlled_circuit_serde_compact() {
+    let pragma_serialization = PragmaControlledCircuit::new(1, Circuit::default());
+    assert_tokens(
+        &pragma_serialization.readable(),
```
Comment (kbarkhqs) [inline roqoqo/tests/integration/operations/pragma_operations.rs:3251]: This one also says `readable()`, should be `compact()`?

**Review by kbarkhqs — CHANGES_REQUESTED** [review]
There are a few comments and parts missing, especially in qoqo

File: qoqo/tests/integration/operations/pragma_operations.rs
Comment (mlodi-hqs) [inline qoqo/tests/integration/operations/pragma_operations.rs:1]: Doesn't it need to be run through the `test_pyo3_remapqubits_error` test?

**Review by mlodi-hqs — COMMENTED** [review]
I've just realized, does the RESERVED_FIELDS constant in `roqoqo-derive/src/lib.rs` need `control_0` and `control_1`, from the 3-qubit operations? I have missed that back in 1.3.

**Comment by kbarkhqs** [comment]
@mlodi-hqs I've added those two fields in a separate pr that should go up today

File: .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs (hunk 1 above)
Comment (nfwvogt) [reply .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs:98]: Fixed

File: .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs (hunk 2 above)
Comment (nfwvogt) [reply .compatibility_tests/compatibility_test_1_4/tests/integration/compatibility_1_4.rs:101]: Added it.

File: qoqo/src/operations/pragma_operations.rs (hunk 3 above)
Comment (nfwvogt) [reply qoqo/src/operations/pragma_operations.rs:1224]: Good catch thanks.

File: qoqo/tests/integration/operations/pragma_operations.rs
Comment (nfwvogt) [reply qoqo/tests/integration/operations/pragma_operations.rs:1]: Added the test

File: roqoqo/tests/integration/operations/pragma_operations.rs (hunk 4 above)
Comment (nfwvogt) [reply roqoqo/tests/integration/operations/pragma_operations.rs:3130]: Added tests.

File: roqoqo/tests/integration/operations/pragma_operations.rs (hunk 5 above)
Comment (nfwvogt) [reply roqoqo/tests/integration/operations/pragma_operations.rs:3251]: Fixed.

File: qoqo/tests/integration/operations/pragma_operations.rs
Comment (nfwvogt) [reply qoqo/tests/integration/operations/pragma_operations.rs:1]: Added test

**Review by kbarkhqs — APPROVED** [review]
